KAFKA-14010: AlterPartition request won't retry when receiving retriable error

showuon opened the pull request:
Jira: [KAFKA-14010](https://issues.apache.org/jira/browse/KAFKA-14010)

When `Partition#submitAlterPartition` submits ISR update and send out the Alter_Isr request, we'll register a callback with `Future.whenComplete` [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/cluster/Partition.scala#L1585). Later, when the Alter_Isr response received, we use `try/finally` to first trigger the callback, and remove the `unsentIsrUpdates` later [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/AlterPartitionManager.scala#L362-L369). Because of this, inside the callback, when we found there is retriable error, we'll retry, but then, we found there's an "in-flight" reqeust because `unsentIsrUpdates` is not removed, which causes the retry didn't actually send out successfully.

In this PR, I did:
1. fix the above described bug by removing the `unsentIsrUpdates` before calling callback. I've confirmed in the callback, we never care about the status of `unsentIsrUpdates`.
~~2. Currently, when we tried to enqueue an ISR update failed due to there's `unsentIsrUpdates`, we throw out `OperationNotAttemptedException`, and later, we'll treat it as no-op from controller, so, we will not retry and also change the state to "committed" [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/cluster/Partition.scala#L1631), which is unexpected. Change it to `IllegalStateException`, so that we'll treat it as "other exception", and will retry this ISR update.~~

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon committed:
KAFKA-14010: fix Alter_Isr can't retry bug

dajac commented on core/src/main/scala/kafka/server/AlterPartitionManager.scala:
```diff
@@ -163,7 +162,7 @@ class DefaultAlterPartitionManager(
       maybePropagateIsrChanges()
     } else {
-      future.completeExceptionally(new OperationNotAttemptedException(
+      future.completeExceptionally(new IllegalStateException(
```
We use `OperationNotAttemptedException` on purpose here. The issue with `IllegalStateException` is that you end up in a tight retry loop.

showuon replied:
OK, thanks.

dajac commented on core/src/main/scala/kafka/server/AlterPartitionManager.scala:
```diff
@@ -359,14 +358,11 @@ class DefaultAlterPartitionManager(
-              } finally {
-                // Regardless of callback outcome, we need to clear from the unsent updates map to unblock further updates
-                unsentIsrUpdates.remove(inflightAlterPartition.topicIdPartition.topicPartition)
+              // we need to clear from the unsent updates map to unblock further updates or retries
```
nit: Should we say something like this?
```
// Regardless of callback outcome, we need to clear from the unsent updates map to unblock further
// updates. We clear it now to allow the callback to submit a new update if needed.
```

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+                                                     isr: List[Int] = List.empty,
+                                                     leaderEpoch: Int = 0,
+                                                     partitionEpoch: Int = 0
+                                                    ): ClientResponse = {
```
nit: Could we format this as follow?
```
def createClientResponseWithAlterPartitionResponse(
  topicPartition: TopicPartition,
  partitionErrorCode: Short,
  isr: List[Int] = List.empty,
  leaderEpoch: Int = 0,
  partitionEpoch: Int = 0
): ClientResponse = {
```

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+  }
+
+  @Test
+  def testPartitionShouldRetryAlterIsrRequest(): Unit = {
```
nit: `AlterPartitionRequest`?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+      createClientResponseWithAlterPartitionResponse(topicPartition, Errors.NONE.code(), List(brokerId, follower1, follower2, follower3), leaderEpoch, partitionEpoch + 1)
+
+    when(mockChannelManager.sendRequest(any(), any()))
+      .thenAnswer(invocation => {
```
nit: `.thenAnswer { invocation =>`?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+      .thenAnswer(invocation => {
+        val controllerRequestCompletionHandler = invocation.getArguments()(1).asInstanceOf[ControllerRequestCompletionHandler]
+        controllerRequestCompletionHandler.onComplete(alterPartitionResponseWithUnknownServerError)
+    })
```
nit: Should this one be aligned on `.thenAnswer`?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
     callback(brokerId, remoteBrokerId, partition)
   }
 
+  def createClientResponseWithAlterPartitionResponse(topicPartition: TopicPartition,
```
nit: Could we make it private?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+        controllerRequestCompletionHandler.onComplete(alterPartitionResponseWithUnknownServerError)
+    })
+      .thenAnswer(invocation => {
+      val controllerRequestCompletionHandler = invocation.getArguments()(1).asInstanceOf[ControllerRequestCompletionHandler]
+      controllerRequestCompletionHandler.onComplete(alterPartitionResponseWithoutError)
+    })
```
nit: Indentation is also weird here.

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+
+    doNothing().when(delayedOperations).checkAndCompleteAll()
+
+    // create a response with error in partition data level, so it'll be handled in AlterPartitionManager, and then Partition callback will re-submit it
```
nit: How about `Fail the first alter partition request with a retryable error to trigger a retry from the partition callback`?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+    val alterPartitionResponseWithUnknownServerError =
+      createClientResponseWithAlterPartitionResponse(topicPartition, Errors.UNKNOWN_SERVER_ERROR.code())
+
+    // create a 2nd response with no error
```
nit: `Complete the ISR expansion`?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+    // Expand ISR
+    fetchFollower(partition, replicaId = follower3, fetchOffset = 10L)
+
+    assertEquals(Set(brokerId, follower1, follower2, follower3), partition.partitionState.isr)
```
nit: Could we also assert the partition epoch?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+    fetchFollower(partition, replicaId = follower3, fetchOffset = 10L)
+
+    assertEquals(Set(brokerId, follower1, follower2, follower3), partition.partitionState.isr)
+    // verify the AlterIsr request will be sent twice
```
nit: `Verify that the AlterPartition request was sent twice

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+    assertEquals(Set(brokerId, follower1, follower2, follower3), partition.partitionState.isr)
+    // verify the AlterIsr request will be sent twice
+    verify(mockChannelManager, times(2)).sendRequest(any(), any())
+    // After retry, the AlterIsr should succeed, and no in-flight request
```
nit: How about `After the retry, the partition state should be committed`?

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+    // After retry, the AlterIsr should succeed, and no in-flight request
+    assertFalse(partition.partitionState.isInflight)
+  }
+
```
It would be great if we could also add a unit test in `AlterPartitionManagerTest` as they issue is in that component.

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,106 @@ class PartitionTest extends AbstractPartitionTest {
+      .setTopicName(topicPartition.topic())
+
+    topicResponse.partitions.add(new AlterPartitionResponseData.PartitionData()
+      .setPartitionIndex(topicPartition.partition())
```
nit: There are a few places where we could omit the parenthesis when calling getters.

dajac reviewed (commented):
@showuon Good catch! I left a few comments.

dajac commented:
@showuon Could you update the title to be inline with the Jira?

showuon committed:
KAFKA-14010: add tests

showuon committed:
KAFKA-14010: address reviewer's comments

showuon commented:
@dajac , thanks for your review! (I was going to ping you, though :)
I've addressed most of your comments. About adding tests in `AlterPartitionManagerTest`, I thought about it, but I can't find a good way to test int `AlterPartitionManagerTest` because the issue is happened during the interaction between `Partition` callback and `AlterPartitionManager` sendRequest callback. If I only tested the `AlterPartitionManagerTest`, it is not meaningful because it indeed removed the `unsentIsrUpdates`. Like the test: `AlterPartitionManagerTest#testUnknownServer` is testing what we expected. 

So, I'm going to leave it as is, unless you have other suggestion. Thanks.

dajac commented:
@showuon I think that we could write a test in `AlterPartitionManagerTest ` that does the following:
* submit an change for foo-0
* add a callback to the future which submit another change for foo-0 when the first request completes
* verify and completes first request
* verify and completes second request

showuon committed:
KAFKA-14010: add test in AlterPartitionManagerTest

showuon commented:
@dajac , thanks for your suggestion. I've added a test: `AlterPartitionManagerTest#testRetryOnPartitionLevelError`. Thanks.

dengziming commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -178,6 +178,46 @@ class AlterPartitionManagerTest {
+    val errorResponse = makeClientResponse(alterPartitionRespWithPartitionError, ApiKeys.ALTER_PARTITION.latestVersion)
+
+    val leaderAndIsr = new LeaderAndIsr(1, 1, List(1,2,3), LeaderRecoveryState.RECOVERED, 10)
+    val callbackCapture: ArgumentCaptor[ControllerRequestCompletionHandler] = ArgumentCaptor.forClass(classOf[ControllerRequestCompletionHandler])
```
nitpick: It's unnecessary to specify the type of local variable.

dengziming reviewed (approved):
This test case is cool, I left a nitpick.

showuon committed:
KAFKA-14010: remove local variable type

dajac commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -178,6 +178,46 @@ class AlterPartitionManagerTest {
   }
 
+  @Test
+  def testRetryOnPartitionLevelError(): Unit = {
```
nit: How about `testSubmitFromCallback`?

dajac commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -178,6 +178,46 @@ class AlterPartitionManagerTest {
+
+    future.whenComplete { (leaderAndIsrResp, e) =>
+      // When entering callback, the `unsentIsrUpdates` element should be removed for possible retry
+      assertFalse(alterPartitionManager.unsentIsrUpdates.containsKey(tp0.topicPartition))
+      assertNull(leaderAndIsrResp)
+      // When receiving retriable error, re-submit the request
+      assertNotNull(e)
+      assertEquals(classOf[UnknownServerException], e.getClass)
```
I ran this test without the fix and I noticed that those assertions, when false, goes nowhere. The only assertion which fails is `verify(brokerToController).sendRequest(any(), callbackCapture.capture())` at L214. There are a big misleading if there are not actionable. Could you double check?

If that is true, an alternative would be to do the following:

```
    val finalFuture = new CompletableFuture[LeaderAndIsr]()
    future.whenComplete { (_, e) =>
      if (e != null) {
        alterPartitionManager.submit(tp0, leaderAndIsr, 0).whenComplete { (result, e) =>
          if (e != null) {
            finalFuture.completeExceptionally(e)
          } else {
            finalFuture.complete(result)
          }
        }
      } else {
        finalFuture.completeExceptionally(new AssertionError("Expected the future to be failed"))
      }
    }
```

and later, we can verify that `finalFuture` contains what we expect: `assertEquals(..., finalFuture.get(100, TimeUnit.MILLISECONDS))`. This also ensures that we get the expected result from the last call to submit.

showuon replied:
Yes, you're right, that's misleading! It's because the future callback will just return the original future exception even though there's new exception thrown from callback, like in the `whenComplete` [java doc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletionStage.html#whenComplete(java.util.function.BiConsumer)) said:

>  this method is not designed to translate completion outcomes, so the supplied action should not throw an exception. However, if it does, the following rules apply: ... if this stage completed exceptionally and the supplied action throws an exception, then the returned stage completes exceptionally with this stage's exception.

I've updated the test as your suggestion. Thank you.

dajac commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -1936,6 +1937,108 @@ class PartitionTest extends AbstractPartitionTest {
+
+    // Fail the first alter partition request with a retryable error to trigger a retry from the partition callback
+    val alterPartitionResponseWithUnknownServerError =
+      createClientResponseWithAlterPartitionResponse(topicPartition, Errors.UNKNOWN_SERVER_ERROR.code())
```
nit: We can remove the `()` here and at L2008.

dajac reviewed (commented):
@showuon Thanks for the update. I left a few small comments.

showuon committed:
KAFKA-14010: refactor test

showuon commented:
@dajac , thanks for the comments. I've updated the PR.

dajac commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -178,6 +179,56 @@ class AlterPartitionManagerTest {
+    val leaderId = 1
+    val leaderEpoch = 1
+    val partitionEpoch = 10
+    val isr = List(1,2,3)
```
nit: We usually add a space after the coma.

dajac commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -178,6 +179,56 @@ class AlterPartitionManagerTest {
+    val leaderEpoch = 1
+    val partitionEpoch = 10
+    val isr = List(1,2,3)
+    val leaderAndIsr = new LeaderAndIsr(leaderId,leaderEpoch, isr, LeaderRecoveryState.RECOVERED, partitionEpoch)
```
nit: A space is missing before `leaderEpoch`.

showuon replied:
nice catch.

dajac commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -545,7 +596,14 @@ class AlterPartitionManagerTest {
-  private def partitionResponse(tp: TopicIdPartition, error: Errors): AlterPartitionResponse = {
+  private def partitionResponse(
+    tp: TopicIdPartition,
+    error: Errors,
```
nit: Should we use `Errors.NONE` as default value?

dajac commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -554,6 +612,10 @@ class AlterPartitionManagerTest {
           .setPartitions(Collections.singletonList(
             new AlterPartitionResponseData.PartitionData()
               .setPartitionIndex(tp.partition())
```
nit: not related to your PR but we could remove the `()` after `partition`.

dajac commented on core/src/test/scala/unit/kafka/server/AlterPartitionManagerTest.scala:
```diff
@@ -178,6 +179,56 @@ class AlterPartitionManagerTest {
+    callbackCapture.getValue.onComplete(retryResponse)
+
+    assertEquals(leaderAndIsr, finalFuture.get(200, TimeUnit.MILLISECONDS))
+    // No more items in unsentIsrUpdates
```
super nit: Comments sometime start with a capital letter and sometime not. I don't have any preferences but I do care about consistency though. Could we use one way or the other?

showuon replied:
I found other tests in this test suite are starting with a capital letter. Following the pattern. Thanks.

dajac reviewed (commented):
@showuon Thanks for the update. I left a few nits.

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-14010

showuon committed:
KAFKA-14010: refactor

dajac reviewed (approved):
LGTM, thanks.

showuon commented:
https://github.com/apache/kafka/pull/12362 is for backporting to 3.2 branch.

ijuma commented:
Is this a regression in 3.2.0? If so, do we know when it was introduced?

dajac commented:
The bug has been there since we introduced the AlterIsr/Partition RPC. https://github.com/apache/kafka/commit/57de67db22eb373f92ec5dd449d317ed2bc8b8d1

ijuma commented:
Thanks @dajac, that's helpful to know.
